File: Textbook_Exercises/Python/Chapter_4/Exercise_4_15/softmax_multiclass_grad_hw.py
# This file is associated with the book
# "Machine Learning Refined", Cambridge University Press, 2016.
# by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

# load the data
def load_data():
    # load data
    data = np.matrix(np.genfromtxt('4class_data.csv', delimiter=','))
    x = np.asarray(data[:,0:2])
    y = np.asarray(data[:,2])
    y.shape = (np.size(y),1)
    return x,y

###### ML Algorithm functions ######
# learn all C separators running gradient descent
def gradient_descent(x,y,alpha):
    # formulate full input data matrix X
    temp = np.shape(x)
    temp = np.ones((temp[0],1))
    X = np.concatenate((temp,x),1)
    X = X.T
    num_classes = len(np.unique(y))
    num_pts = len(y)
    W = np.random.randn(3,num_classes)

    # record objective value at each iteration to check that algorithm is working properly
    obj_path = []
    obj = calculate_obj(X,y,W)
    obj_path.append(obj)

    # gradient descent loop
    iter = 1
    max_its = 1000
    while iter < max_its:

        # calculate gradient
# --->  grad = 

        # full gradient completely calculated - take gradient step
        W = W - alpha*grad
        iter+= 1

        # update path containers - used to check that algorithm is working
        obj = calculate_obj(X,y,W)
        obj_path.append(obj)

    # plot objective value at each iteration to make sure everything works properly
    obj_path = np.asarray(obj_path)
    obj_path.shape = (max_its,1)
    # plt.plot(np.asarray(obj_path))

    return W

# calculate the objective value for a given input weight w
def calculate_obj(X,y,W):
    # loop for cost function
    cost = 0
    for p in range(0,len(y)):
        s = int(y[p])-1
        p_temp = 0
        for j in range(0,len(np.unique(y))-1):
            p_temp += np.exp(np.dot(X[:,p].T,(W[:,j] - W[:,s])))
        p_temp = np.log(p_temp)

        # update cost
        cost+=p_temp
    return cost

# plot data, separators, and fused fule
def plot_all(x,y,W):
    # initialize figure, plot data, and dress up panels with axes labels etc.
    num_classes = np.size(np.unique(y))
    color_opts = np.array([[1,0,0.4], [ 0, 0.4, 1],[0, 1, 0.5],[1, 0.7, 0.5],[0.7, 0.6, 0.5]])
    f,axs = plt.subplots(1,3,facecolor = 'white')
    for a in range(0,3):
        for i in range(0,num_classes):
            s = np.argwhere(y == i+1)
            s = s[:,0]
            axs[a].scatter(x[s,0],x[s,1], s = 30,color = color_opts[i,:])

        # dress panel correctly
        axs[a].set_xlabel('$x_1$',fontsize=20,labelpad = 20)
        axs[a].set_ylabel('$x_2$',fontsize=20,rotation = 0,labelpad = 20)
        axs[a].set_xlim(0,1)
        axs[a].set_ylim(0,1)
        axs[a].set(aspect = 'equal')

    r = np.linspace(0,1,150)
    for i in range(0,num_classes):
        z = -W[0,i]/W[2,i] - W[1,i]/W[2,i]*r
        axs[1].plot(r,z,'-k',linewidth = 2,color = color_opts[i,:])

    # fuse individual subproblem separators into one joint rule
    r = np.linspace(0,1,300)
    s,t = np.meshgrid(r,r)
    s = np.reshape(s,(np.size(s),1))
    t = np.reshape(t,(np.size(t),1))
    h = np.concatenate((np.ones((np.size(s),1)),s,t),1)
    f = np.dot(W.T,h.T)
    z = np.argmax(f,0)
    f.shape = (np.size(f),1)
    s.shape = (np.size(r),np.size(r))
    t.shape = (np.size(r),np.size(r))
    z.shape = (np.size(r),np.size(r))

    for i in range(0,num_classes + 1):
        axs[2].contour(s,t,z,num_classes-1,colors = 'k',linewidths = 2.25)

# main wrapper
def softmax_multiclass_grad_hw():
    # load data
    x,y = load_data()

    # perform gradient descent on softmax multiclass
    alpha = 10**(-2)    # step length, tune to your heat's desire!
    W = gradient_descent(x,y,alpha)           # learn all C vs notC separators
    plot_all(x,y,W)
    plt.show()

softmax_multiclass_grad_hw()

File: mlrefined_exercises/ed_1/MATLAB/Chapter_4/Exercise_4_15/softmax_multiclass_grad_hw.m
function softmax_multiclass_grad_hw()

% This file is associated with the book
% "Machine Learning Refined", Cambridge University Press, 2016.
% by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

% load in data
[X,y] = load_data();


% initializations
N = size(X,2);
C = length(unique(y));
X = [ones(size(X,1),1), X]';
W0 = randn(N+1,C);
alpha = 0.1;


% find separators via multiclass softmax classifier
W = softmax_multiclass_grad(X,y,W0,alpha);


% plot the separators as well as final classification
plot_separators(W, X, y)


%%%%%%%%%%%%%%%%% subfunctions %%%%%%%%%%%%%%%%%%%
function W = softmax_multiclass_grad(X,y,W0,alpha)
  
    % initialize
    max_its = 10000; 
    [N,P] = size(X);
    C = length(unique(y));
    W = W0;
    k = 1;

    %%% main %%%
    while k <= max_its
   

% ----> grad = 
        W = W - alpha*grad;
        
        % update counter
        k = k + 1;
        
    end
end

function [X,y] = load_data()
    data = csvread('4class_data.csv');
    X = data(:,1:end - 1);
    y = data(:,end);       
end

function plot_separators(W,X,y,deg)
    
    red =  [ 1 0 .4];
    blue =  [ 0 .4 1];
    green = [0 1 0.5];
    cyan = [1 0.7 0.5];
    grey = [.7 .6 .5];
    colors = [red;blue;green;cyan;grey]; 
    % plot data
    subplot(1,3,1)
    plot_data(X,y)
    subplot(1,3,2)
    plot_data(X,y)
    subplot(1,3,3)
    plot_data(X,y)

    %%% plot all linear separators %%%
    subplot(1,3,2)
    num_classes = length(unique(y));
    x = [0:0.01:1];
    for j = 1:num_classes
        hold on
        w = W(:,j);    
        plot (x,(-w(1)-w(2)*x)/w(3),'Color',colors(j,:),'linewidth',2);
    end

    %%% generate max-separator surface %%%
    s = [0:0.005:1];
    [s1,s2] = meshgrid(s,s);
    s1 = reshape(s1,numel(s1),1);
    s2 = reshape(s2,numel(s2),1);
    
    % compute criteria for each point in the range [0,1] x [0,1]
    square = [s1(:), s2(:)];
    p = [ones(size(s1(:),1),1),s1(:), s2(:)];
    f = W'*p';         
    [f,z] = max(f,[],1);
    
    % fill in appropriate regions with class colors
    subplot(1,3,3)
    ind = find(z == 1);
    k = boundary(square(ind,:));
    v = [square(ind(k),:), ones(length(k),1)];
    f = 1:length(k);
    patch('Faces',f,'Vertices',v,'FaceColor',red,'FaceAlpha',0.3)

    ind = find(z == 2);
    k = boundary(square(ind,:));
    v = [square(ind(k),:), 2*ones(length(k),1)];
    f = 1:length(k);
    patch('Faces',f,'Vertices',v,'FaceColor',blue,'FaceAlpha',0.3)
    
    ind = find(z == 3);
    k = boundary(square(ind,:));
    v = [square(ind(k),:), 3*ones(length(k),1)];
    f = 1:length(k);
    patch('Faces',f,'Vertices',v,'FaceColor',green,'FaceAlpha',0.3)
    
    ind = find(z == 4);
    k = boundary(square(ind,:));
    v = [square(ind(k),:), 4*ones(length(k),1)];
    f = 1:length(k);
    patch('Faces',f,'Vertices',v,'FaceColor',cyan,'FaceAlpha',0.3)
    
    ind = find(z == 5);
    k = boundary(square(ind,:));
    v = [square(ind(k),:), 4*ones(length(k),1)];
    f = 1:length(k);
    patch('Faces',f,'Vertices',v,'FaceColor',grey,'FaceAlpha',0.3)
    
    % produce decision boundary
    s1 = reshape(s1,[length(s),length(s)]);
    s2 = reshape(s2,[length(s),length(s)]);   
    z = reshape(z,[length(s),length(s)]);   
    num_classes = length(unique(z));
    subplot(1,3,3)
    for i = 1:num_classes - 1
       hold on
       contour(s1,s2,z,[i + 0.5,i + 0.5],'Color','k','LineWidth',2)
    end
    
    % make plot real nice lookin'
    for i = 1:3
        subplot(1,3,i)
        axis([0 1 0 1])
        axis square
        xlabel('x_1','FontName','cmmi9','Fontsize',18)
        ylabel('x_2','FontName','cmmi9','Fontsize',18)
        set(get(gca,'YLabel'),'Rotation',0)
        zlabel('y','FontName','cmmi9','Fontsize',18)
        set(get(gca,'ZLabel'),'Rotation',0)

        set(gca,'XTick',[0,1])
        set(gca,'YTick',[0,1])
        set(gca,'ZTick',[0:1:num_classes])
        set(gcf,'color','w');
    end
end

function plot_data(X,y)
    red =  [ 1 0 .4];
    blue =  [ 0 .4 1];
    green = [0 1 0.5];
    cyan = [1 0.7 0.5];
    grey = [.7 .6 .5];
    colors = [red;blue;green;cyan;grey]; 
    % how many classes in the data? maximum 4 here.
    class_labels = unique(y);           % class labels
    num_classes = length(class_labels);

    % plot data
    for i = 1:num_classes
        class = class_labels(i);
        ind = find(y == class);
        hold on
        scatter3(X(2,ind),X(3,ind),class*ones(length(ind),1),'Linewidth',2,'Markeredgecolor',colors(i,:),'markerFacecolor','none');
        hold on
        scatter3(X(2,ind),X(3,ind),class*ones(length(ind),1),'Linewidth',2,'Markeredgecolor',colors(i,:),'markerFacecolor','none');
    end
    axis([0 1 0 1])
    axis square
    box on
end
end